### PR TITLE
TEP-0012: API Spec

### PR DESCRIPTION
**Pull request by ImJasonH**
[Rendered here](https://github.com/ImJasonH/community/blob/api-spec/teps/0008-api-spec.md)

/assign @bobcatfish 
/assign @vdemeester 
/assign @afrittoli

### CODE REVIEW

File: teps/0008-api-spec.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+The API today is large, sprawling, and relatively fast-moving. This is a _good thing_, as it provides users a plethora of options to describe and run their workloads. But it's also a challenge for new users, and for potential platform providers and implementers. Inclusion of Kubernetes built-in types (`Container`, `PodSpec`, `Volume`, etc.) accounts for the majority of the API's total surface area, even though many of those options are rarely used by users, and may not even be supported on most platforms. Some fields are ignored outright by the Kubernetes-based reference implementation (`imagePullPolicy`), and some are unusable in practice (`livenessProbe`, `terminationMessagePath`). Support for these fields is difficult to reason about in the abstract, particularly in cases where a future implementation might not be running on Kubernetes at all.
```
Comment (bobcatfish) [inline teps/0008-api-spec.md:38]: 👍 👍 👍

File: teps/0008-api-spec.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+This will serve as documentation to end users and client authors (including Tekton's own CLI and Dashboard) about which features it should expect to be supported.
+
+Document the lifecycle of a Tekton Pipelines API feature, through alpha to beta to GA, and possibly separately to being required to satisfy API spec conformance. This process should cover how to communicate to clients and existing implementations that a new field will newly be considered required.
+
+Produce and maintain a conformance test suite which can be run against a candidate Tekton implementation to determine whether it satisfies the documented spec. This test suite should be regularly run against a platform to guard against regressions.
+
+### Non-Goals
+
+Document the [runtime contract](https://tekton.dev/docs/pipelines/container-contract/), or prescribe specific implementations of supporting services such as access control, observability and resource management.
```
Comment (bobcatfish) [inline teps/0008-api-spec.md:54]: i would think that conformance tests would be testing some parts of the runtime contract as well - i assume one of the goals of conformance (maybe we could add above if you agree) is to ensure portability b/w implementations (or maybe not?)

or maybe this is something we'd handle in a TEP specific to conformance and you just want to focus on the API spec here?

Comment (bobcatfish) [reply teps/0008-api-spec.md:54]: o yeah you mentioned portability in the user stories below, so maybe it makes sense to include

File: teps/0008-api-spec.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+The first iteration of the API spec should require the bare minimum functionality from a Tekton implementation -- that is, `TaskRun`s can be specified with `TaskSpec`s defined inline to specify `steps`. In addition to `steps`, a minimal `TaskRun` should support `params` and `results`, `workspaces`, and associated `status` fields. (The full list will be proposed in the initial draft API spec).
+
+Task references, including `ClusterTask`s and OCI image bundles, will not be included in the initial version, nor will `Pipeline`s or `PipelineRun`s, `Condition`s, etc. -- these will be added as appropriate in later iterations.
+
+This initial iteration will allow us to focus on the conformance test suite, and quickly iterate by expanding the API spec requirements and recommendations.
+
+### User Stories (optional)
+
+#### End User
```
Comment (bobcatfish) [inline teps/0008-api-spec.md:74]: side note: we should update the list at https://github.com/tektoncd/community/blob/master/user-profiles.md (i like your list) (probably just needs an update)

File: teps/0008-api-spec.md [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+A second or third API implementation may never materialize, and the work to document and enforce conformance may be wasted. By limiting initial requirements we can both lower the barrier to entry for future implementations, and limit the amount of work we expend on the conformance test suite, until such a time as the benefit of future implementations makes further investment necessary.
+
+## Test Plan
+
+We should add a separate test suite to the Tekton Pipelines repo that sends requests to a specified Tekton Pipelines implementation and expects that the request is accepted and that expected behavior is observed as a result.
+
+For example, sending a request to create a `TaskRun` to the API that only specifies required fields should succeed against any conformant implementation. Periodically polling that `TaskRun` resource until it indicates it completed successfully (or otherwise) should also succeed against any conformant implementation.
+
+To be clear, this test suite doesn't need to focus as much on _correctness_ -- that's what existing end-to-end tests cover -- but should only send simple requests and expect to observe conformant behavior.
```
Comment (bobcatfish) [inline teps/0008-api-spec.md:100]: am i right to assume we'd elaborate on this in a future conformance tep?

Comment (ImJasonH) [reply teps/0008-api-spec.md:100]: I think we'd elaborate on the specifics of this in the PR(s) that add the conformance tests.

File: teps/0008-api-spec.md (hunk 2 above)
Comment (ImJasonH) [reply teps/0008-api-spec.md:54]: I'd like this to be the last TEP for now related to API specification -- the next ones will come as we expand its coverage to include Tasks/Pipelines/PipelineRuns.

This TEP is tldr "we should document and test conformance", if there's something you think we should cover that's not roughly outlined in this TEP, we should push on that in this TEP I think.

Testing the runtime contract is going to be quite a bit more work, I think, and I think it's unlikely that spec-covered features ("run this image with these args") will have divergent behavior between two implementations. If we find that's not the case, then we can document the runtime contract better, and even add it to conformance tests. But I'd like to figure out if it's a problem before we commit to testing the entire runtime contract right out the gate.

wdyt?

File: teps/0008-api-spec.md [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+## Requirements
+
+Acknowledgement and approval by Tekton's governance board, and stakeholders in the CLI and Dashboard projects that an API spec will be forthcoming. Discussion about specifics of the API spec can happen in the PR that proposes it.
+
+## Proposal
+
+The first iteration of the API spec should require the bare minimum functionality from a Tekton implementation -- that is, `TaskRun`s can be specified with `TaskSpec`s defined inline to specify `steps`. In addition to `steps`, a minimal `TaskRun` should support `params` and `results`, `workspaces`, and associated `status` fields. (The full list will be proposed in the initial draft API spec).
+
+Task references, including `ClusterTask`s and OCI image bundles, will not be included in the initial version, nor will `Pipeline`s or `PipelineRun`s, `Condition`s, etc. -- these will be added as appropriate in later iterations.
```
Comment (dlorenc) [inline teps/0008-api-spec.md:68]: Omitting references is a pretty large gap - it essentially means most of the catalog can't be used. Should we provide a catalog of inline taskruns?

Comment (ImJasonH) [reply teps/0008-api-spec.md:68]: This is a temporary limitation, to reduce the scope of the initial spec surface. We absolutely fully intend to suppot the catalog in the fullness of time.

I think there's a discussion we could have about whether platforms _must_ support it with in-service Task/Pipeline CRUD or whether they _may_ support it only via OCI bundles (#137), and tbh I hope having that discussion outside the initial spec proposal (i.e., "should we _have_ a spec?") makes this much easier to accept.

**Review by vdemeester — APPROVED** [review]
/meow

**Comment by bobcatfish** [comment]
Discussed [in the api working group today](https://docs.google.com/document/d/1rPR7m1Oj0ip3bpd_bcS1sjZyPgGi_g9asF5YrExeESc/edit#heading=h.wqdxc9tatgb0) and felt that it was OK for this to go straight to implementable, esp since implementing this means creating another TEP with specific API details.

/lgtm

**Comment by 06kellyjac** [comment]
is there an issue to track implementation?

**Comment by vdemeester** [comment]
@06kellyjac there is not 😅 
/cc @imjasonh @jerop

**Pull request merged**